packages/ifpack2/src/Ifpack2_Details_MultiVectorLocalGatherScatter.hpp

```diff
@@ -101,6 +101,7 @@ class MultiVectorLocalGatherScatter {
     using Teuchos::ArrayRCP;
     const size_t numRows = X_out.getLocalLength ();
     const size_t numVecs = X_in.getNumVectors ();
+    Kokkos::fence();
     for (size_t j = 0; j < numVecs; ++j) {
       ArrayRCP<const InScalar> X_in_j = X_in.getData(j);
       ArrayRCP<OutScalar> X_out_j = X_out.getDataNonConst(j);
```

Comment:
Should there be a ``X_out.modify_host()`` at the end of each of these?

I'm also in favor of moving these to device, or even better executing it wherever X_in is most up to date like idot does.

Reply:
I updated with modify_host calls for the MV functions (understanding we will probably refactor anyways).